sirepo/template/ml.py

```diff
@@ -287,13 +287,83 @@ def stateful_compute_compute_column_info(data):
     return _compute_column_info(data.dataFile)
 
 
+def stateless_compute_get_remote_data(data):
+    return _get_remote_data(data.url, data.headers_only)
+
+
+def stateless_compute_remote_data_bytes_loaded(data):
+    return _remote_data_bytes_loaded(data.filename)
+
+
 def write_parameters(data, run_dir, is_parallel):
     pkio.write_text(
         run_dir.join(template_common.PARAMETERS_PYTHON_FILE),
         _generate_parameters_file(data),
     )
 
 
+def _build_model_py(v):
+    def _import_layers(v):
+        return "".join(", " + n for n in v.layerImplementationNames)
+
+    def _conv_args(layer):
+        if layer.layer not in ("Conv2D", "Transpose", "SeparableConv2D"):
+            return
+        return f"""{layer.dimensionality},
+    activation="{layer.activation}",
+    kernel_size=({layer.kernel}, {layer.kernel}),
+    strides={layer.strides},
+    padding="{layer.padding}"
+    """
+
+    def _pooling_args(layer):
+        return f'''pool_size=({layer.size}, {layer.size}),
+    strides={layer.strides},
+    padding="{layer.padding}"'''
+
+    args_map = PKDict(
+        Activation=lambda layer: f'"{layer.activation}"',
+        AlphaDropout=lambda layer: layer.dropoutRate,
+        AveragePooling2D=lambda layer: _pooling_args(layer),
+        BatchNormalization=lambda layer: f"momentum={layer.momentum}",
+        Conv2D=lambda layer: _conv_args(layer),
+        Dense=lambda layer: f'{layer.dimensionality}, activation="{layer.activation}"',
+        Dropout=lambda layer: layer.dropoutRate,
+        Flatten=lambda layer: "",
+        GaussianDropout=lambda layer: layer.dropoutRate,
+        GaussianNoise=lambda layer: layer.stddev,
+        GlobalAveragePooling2D=lambda layer: "",
+        MaxPooling2D=lambda layer: _pooling_args(layer),
+        SeparableConv2D=lambda layer: _conv_args(layer),
+        Conv2DTranspose=lambda layer: _conv_args(layer),
+        UpSampling2D=lambda layer: f'size={layer.size}, interpolation="{layer.interpolation}"',
+        ZeroPadding2D=lambda layer: f"padding=({layer.padding}, {layer.padding})",
+    )
+
+    def _layer_args(layer):
+        assert layer.layer in args_map, ValueError(f"invalid layer.layer={layer.layer}")
```

Comment:
I don't think _layer_args is necessary. Returning ValueError on assert doesn't make sense. Just let it naturally fail fast.

Reply:
[0ca67b9](https://github.com/radiasoft/sirepo/pull/4671/commits/0ca67b9604d6873b11bb2249fbd9a6565dca7226)